src/common/container_helpers.h

```diff
@@ -0,0 +1,170 @@
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Miscellaneous container helpers.
+
+#pragma once
+
+//local headers
+
+//third party headers
+
+//standard headers
+#include <algorithm>
+#include <unordered_map>
+#include <utility>
+
+//forward declarations
+
+
+namespace tools
+{
+
+/// convert an arbitrary function to functor
+template <typename F>
+inline auto as_functor(F f)
+{
+    return [f = std::move(f)](auto&&... args) { return f(std::forward<decltype(args)>(args)...); };
+}
+/// convert a binary comparison function to a functor
+/// note: for most use-cases 'const T&' will work because only non-trivial types need a user-defined comparison operation
+template <typename T, typename ComparisonOpT = bool(const T&, const T&)>
+inline auto compare_func(ComparisonOpT comparison_op_func)
+{
+    static_assert(
+            std::is_same<
+                    bool,
+                    decltype(
+                        comparison_op_func(
+                            std::declval<std::remove_cv_t<T>>(),
+                            std::declval<std::remove_cv_t<T>>()
+                        )
+                    )
+                >::value,
+            "invalid callable - expected callable in form bool(T, T)"
+        );
+
+    return as_functor(std::move(comparison_op_func));
+}
+/// test if a container is sorted and unique according to a comparison criteria (defaults to operator<)
+/// NOTE: ComparisonOpT must establish 'strict weak ordering' https://en.cppreference.com/w/cpp/named_req/Compare
+template <typename T, typename ComparisonOpT = std::less<typename T::value_type>>
+bool is_sorted_and_unique(const T &container, ComparisonOpT comparison_op = ComparisonOpT{})
+{
+    using ValueT = typename T::value_type;
+    static_assert(
+            std::is_same<
+                    bool,
+                    decltype(
+                        comparison_op(
+                            std::declval<std::remove_cv_t<ValueT>>(),
+                            std::declval<std::remove_cv_t<ValueT>>()
+                        )
+                    )
+                >::value,
+            "invalid callable - expected callable in form bool(ValueT, ValueT)"
+        );
+
+    if (!std::is_sorted(container.begin(), container.end(), comparison_op))
+        return false;
+
+    if (std::adjacent_find(container.begin(),
```

Comment:
If the container is sorted, `std::lower_bound` instead? If the list is large, `lower_bound` will make a difference since its a logarithmic search.

Reply:
Could've been solved by me reading `adjacent_find` documentation.